src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/Cohost/CohostSemanticTokensRangeEndpoint.cs

```diff
@@ -16,6 +16,7 @@
 
 namespace Microsoft.VisualStudio.LanguageServerClient.Razor.Cohost;
 
+[Shared]
```

Comment:
>  Razor has any scenarios where MEF exports aren't intended to be singletons.

I don't know if this is relevant, but the VS LSP client, as far as I'm aware, creates instances of `ILanguageClient` when the first file of a relevant content type is opened, and disposes them when the solution is closed. I don't know if that would flow down to these endpoints having the same behaviour, or if something in Roslyn and/or something I did in the Roslyn cohost bits rule that out, but thought I'd mention it here.

I should mention I don't think this is necessarily something we need to have, as at worst the only state our endpoints should have would be knowledge of the clients capabilities, and the client would requery for that, and hence we would re-set that state, with every creation.

Reply:
Thanks for mentioning that. I'll dig in to our language client while thinking through our MEF exports and layering to see if singletons make sense. I think they probably still do, since they should be stateless and the services that they use are likely all singletons, but I'll make sure. Thanks!